File: Depth-First Search/Binary tree/Searching/Easy/LET_no270_Closest Binary Search Tree Value(iterative).py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def closestValue(self, root: Optional[TreeNode], target: float) -> int:
        upper = lower = root
        while root:
            if root.val > target:
                upper = root
                root = root.left
            elif root.val < target:
                lower = root
                root = root.right
            else:
                return root.val
            
        if abs(upper.val - target) < abs(lower.val - target):
            return upper.val
        return lower.val

File: Depth-First Search/Binary tree/Searching/Easy/LET_no270_Closest Binary Search Tree Value(iterative).cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int closestValue(TreeNode* root, double target) {
        TreeNode *lower = root, *upper = root;
        
        while (root) {
            if (root->val > target) {
                upper = root;
                root = root->left;
            }
            else if (root->val < target) {
                lower = root;
                root = root->right;
            }
            else 
                return root->val;
        }
        
        return abs(upper->val - target) < abs(lower->val - target) ? upper->val : lower->val;
    }
};